### PR TITLE
Typesense connector

### PR DESCRIPTION
**Pull request by LukBed**
New connector for Typesense, https://typesense.org/

### CODE REVIEW

File: typesense/src/test/resources/application.conf [hunk 1]
```diff
@@ -0,0 +1,5 @@
+akka {
```
Comment (lbialy) [inline typesense/src/test/resources/application.conf:1]: not sure if this should be reference.conf as it's test scope but it's worth checking

**Review by lbialy — CHANGES_REQUESTED** [review]
More like - request for comments. Some decisions are not obvious, I'd like to learn more. Most of the PR LGTM though, java interop looks nice!

File: typesense/src/main/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,10 @@
+alpakka.typesense {
```
[the remark this answers is not in the document]
Comment (LukBed) [reply typesense/src/main/resources/application.conf:1]: I copied it from reference module when it was called application.conf, but I see in all modules it's called reference.conf - so I'll rename it

File: typesense/src/test/resources/application.conf (hunk 1 above)
Comment (LukBed) [reply typesense/src/test/resources/application.conf:1]: To be honest I don't know, I see in most modules it's called reference in main and application in test

File: typesense/src/main/scala/akka/stream/alpakka/typesense/documents.scala [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+sealed abstract class IndexDocumentAction
+
+object IndexDocumentAction {
+  sealed abstract class Create extends IndexDocumentAction
+  case object Create extends Create
+
+  /**
+   * Java API
+   */
+  def create: Create = Create
```
[the remark this answers is not in the document]
Comment (LukBed) [reply typesense/src/main/scala/akka/stream/alpakka/typesense/documents.scala:96]: Fixed

File: typesense/src/main/scala/akka/stream/alpakka/typesense/TypesenseResult.scala [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+import akka.annotation.InternalApi
+import akka.http.scaladsl.model.StatusCode
+
+sealed trait TypesenseResult[+T] {
+  def map[S](f: T => S): TypesenseResult[S]
+  def get(): Option[T]
+}
+
+final class SuccessTypesenseResult[T] @InternalApi private[typesense] (val value: T) extends TypesenseResult[T] {
```
[the remark this answers is not in the document]
Comment (LukBed) [reply typesense/src/main/scala/akka/stream/alpakka/typesense/TypesenseResult.scala:15]: For binary compatibility based on contributor guide
https://github.com/akka/alpakka/blob/master/contributor-advice.md

> All Akka APIs aim to evolve in a binary compatible way within minor versions.
> 1. Do not use any default arguments
> 2. Do not use case classes (as the public copy method relies on default arguments)
> 3. To generate a case class replacement, consider using Kaze Class

File: typesense/src/main/scala/akka/stream/alpakka/typesense/TypesenseSettings.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ */
+
+package akka.stream.alpakka.typesense
+
+import akka.annotation.InternalApi
+import scala.jdk.DurationConverters._
+
+import scala.concurrent.duration.FiniteDuration
+
+final class TypesenseSettings @InternalApi private[typesense] (val host: String,
```
[the remark this answers is not in the document]
Comment (LukBed) [reply typesense/src/main/scala/akka/stream/alpakka/typesense/TypesenseSettings.scala:12]: As above - for binary compatibility

File: typesense/src/main/scala/akka/stream/alpakka/typesense/TypesenseSettings.scala [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+object TypesenseSettings {
+  def apply(host: String, apiKey: String, retrySettings: RetrySettings): TypesenseSettings =
+    new TypesenseSettings(host, apiKey, retrySettings)
+
+  def create(host: String, apiKey: String, retrySettings: RetrySettings): TypesenseSettings =
+    new TypesenseSettings(host, apiKey, retrySettings)
+}
+
+final class RetrySettings @InternalApi private[typesense] (val maxRetries: Int,
```
[the remark this answers is not in the document]
Comment (LukBed) [reply typesense/src/main/scala/akka/stream/alpakka/typesense/TypesenseSettings.scala:44]: As above - for binary compatibility

File: typesense/src/main/scala/akka/stream/alpakka/typesense/documents.scala [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+package akka.stream.alpakka.typesense
+
+import akka.annotation.InternalApi
+import akka.stream.alpakka.typesense.javadsl.FilterDeleteDocumentsQueryDsl
+
+import java.util.OptionalInt
+import scala.compat.java8.OptionConverters.RichOptionalInt
+import scala.jdk.CollectionConverters.{CollectionHasAsScala, SeqHasAsJava}
+
+final class IndexDocument[T] @InternalApi private[typesense] (val collectionName: String,
```
[the remark this answers is not in the document]
Comment (LukBed) [reply typesense/src/main/scala/akka/stream/alpakka/typesense/documents.scala:14]: As above - for binary compatibility

File: typesense/src/main/scala/akka/stream/alpakka/typesense/impl/TypesenseHttp.scala [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  TypesenseSettings
+}
+import spray.json.{JsValue, JsonReader}
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Success, Try}
+
+@InternalApi private[typesense] object TypesenseHttp {
+  object PrepareRequestEntity {
+    def json(body: JsValue): RequestEntity = HttpEntity(ContentTypes.`application/json`, body.prettyPrint)
```
[the remark this answers is not in the document]
Comment (LukBed) [reply typesense/src/main/scala/akka/stream/alpakka/typesense/impl/TypesenseHttp.scala:34]: Fixed

File: typesense/src/main/scala/akka/stream/alpakka/typesense/TypesenseResult.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (lbialy) [reply typesense/src/main/scala/akka/stream/alpakka/typesense/TypesenseResult.scala:15]: aaaah, cool, TIL!

File: typesense/src/test/resources/application.conf (hunk 1 above)
Comment (lbialy) [reply typesense/src/test/resources/application.conf:1]: it makes sense to some degree, reference in main as all reference.conf files are merged and compacted to provide defaults and application.conf is the configuration that overrides defaults afaik

File: typesense/src/main/resources/reference.conf [hunk 9]
```diff
@@ -0,0 +1,10 @@
+alpakka.typesense {
+  host = "http:localhost:8108"
+  api-key = "Hu52dwsas2AdxdE"
```
Comment (ennru) [inline typesense/src/main/resources/reference.conf:3]: These values should be left blank to force users to set them instead.
Add `src/test/resources/application.conf` to set these for testing.

File: typesense/src/main/scala/akka/stream/alpakka/typesense/TypesenseResult.scala [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  override def map[S](f: T => S): TypesenseResult[S] = new SuccessTypesenseResult(f(value))
+
+  override def get(): Option[T] = Some(value)
```
Comment (ennru) [inline typesense/src/main/scala/akka/stream/alpakka/typesense/TypesenseResult.scala:18]: Should this be a `val`?
Could `value` be `null`? What would it mean?

File: typesense/src/test/resources/application.conf (hunk 1 above)
Comment (ennru) [reply typesense/src/test/resources/application.conf:1]: `application.conf` is correct. This is the use-side of the configuration.

**Review by ennru — COMMENTED** [review]
Looks great, so far I got at a first look.
But lacks documentation other than code samples.

**Comment by ennru** [comment]
...and it needs to be added to the CI build.

**Comment by LukBed** [comment]
I'm closing PR and I don't transfer rights within CLA

**Pull request closed without merging**